src/wp-includes/global-styles-and-settings.php

```diff
@@ -255,3 +255,70 @@ function ( $item ) {
 		}
 	}
 }
+
+/**
+ * Whether a theme or its parent have a theme.json file.
+ *
+ * @since 6.2.0
+ *
+ * @return bool
+ */
+function wp_theme_has_theme_json() {
+	/*
+	 * By using the 'theme_json' group, this data is marked to be non-persistent across requests.
+	 * See `wp_cache_add_non_persistent_groups` in src/wp-includes/load.php and other places.
+	 *
+	 * The rationale for this is to make sure derived data from theme.json
+	 * is always fresh from the potential modifications done via hooks
+	 * that can use dynamic data (modify the stylesheet depending on some option,
+	 * settings depending on user permissions, etc.).
+	 * See some of the existing hooks to modify theme.json behaviour:
+	 * https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/
+	 *
+	 * A different alternative considered was to invalidate the cache upon certain
+	 * events such as options add/update/delete, user meta, etc.
+	 * It was judged not enough, hence this approach.
+	 * See https://github.com/WordPress/gutenberg/pull/45372
+	 */
+	$cache_group       = 'theme_json';
+	$cache_key         = 'wp_theme_has_theme_json';
+	$theme_has_support = wp_cache_get( $cache_key, $cache_group );
+
+	/*
+	 * $theme_has_support is stored as an int in the cache.
+	 *
+	 * The reason not to store it as a boolean is to avoid working
+	 * with the $found parameter which apparently had some issues in some implementations
+	 * https://developer.wordpress.org/reference/functions/wp_cache_get/
+	 *
+	 * Ignore cache when `WP_DEBUG` is enabled, so it doesn't interfere with the theme developers workflow.
+	 */
+	if ( ! WP_DEBUG && is_int( $theme_has_support ) ) {
```

Comment:
This is the one thing here I don't think we should be doing. I realize we have had multiple discussions around it on the Gutenberg repo. And while this is currently in Gutenberg `trunk`, merging it here would set it as a new precedent (IMO a bad precedent) in core, so I would like to avoid it.

I would see two options here:
1. Either we remove the `! WP_DEBUG` bit completely for now.
2. Or we come up with an appropriate solution that allows being configured without having to rely _only_ on `WP_DEBUG` (e.g. what I suggested in https://github.com/WordPress/gutenberg/issues/45912).

Since at this point we're only using a non-persistent cache anyway, having this check is not as critical, so I would be okay going with option 1 for now.

Reply:
Hey @felixarntz, @oandregal is away this week. To move this forward, I'll update this PR for option 1. And then if option 2 becomes available, then a separate PR can be done for it.

Reply:
Hmm scanning Core, there are multiple instances of `WP_DEBUG` being used in conditionals including:

- Global Styles: whether to use cache in `wp_get_global_stylesheet()` (introduced in 5.9.0). 
- Global Styles: whether to use the cache for `global_styles_svg_filters` (introduced in 5.9.1)
- In REST API: whether to bail out  for handling deprecations (in `rest_handle_deprecated_function()`) (introduced in 4.4.0).
- In `wp_version_check()` for determine which error level to populate in a notification.
- In `render_block_core_block()` for deciding whether to display an error notification.
- In the Upgrader for `''no signature could be found''` handling in `WP_Upgrader::run()` (introduced in 2.8.0 r45262).

and others.

Since there is precedence, @felixarntz are you okay with keeping `WP_DEBUG` here?

Reply:
@hellofromtonya All of those are different though. The first two are a bit related, and to be fair, had I reviewed that code I would have pointed out a similar thing. But still the conditions there at least consider a few other options as well - which I still think shows that we should have a centralized way to indicate whether or not we want block theme related logic to be cached (see https://github.com/WordPress/gutenberg/issues/45912).

The other for are different as they aren't related to block theme caching. My point here is that we shouldn't use `WP_DEBUG` to decide whether we cache block theme related logic, as that isn't really about debugging as much as it is about being able to develop a theme and see changes reflected immediately (by disabling cache).

But in any case, since for now those are non-persistent caches, I don't even think at this point we need any check here, since those values are only cached within a single request, not across requests. So I would still advise to remove it at this point and iterate with a more appropriate solution if/once eventually we enhance the caching to be persistent across requests.

Reply:
Uh, already replied at https://github.com/WordPress/wordpress-develop/pull/3712/files#r1071617204, should have looked here first :) 

The intention here is to disable caching when developing. You're right that `WP_DEBUG` is not directly related to WordPress being run in a "development mode", but is pretty close. You would not/should not turn debugging on in production, right? 

> those values are only cached within a single request, not across requests

Caching is caching is... caching :) 
If there are chances it may interfere with (theme) development, it should be turned off. There are sooo many examples of "weird", hard to find bugs caused by not invalidating caches. Would be quite better to err on the side of safety.

As I mentioned in the [other comment](https://github.com/WordPress/wordpress-develop/pull/3712/files#r1071617204), seems WP needs a way (another constant) to signify "development mode". Once that is added, these uses of `WP_DEBUG` can be removed.

Reply:
@azaozz I like your notion of "development mode". My suggestion had been something like https://github.com/WordPress/gutenberg/issues/45912, but yours is more generally applicable.

> If there are chances it may interfere with (theme) development, it should be turned off.

Caching within a single request does not interfere with theme development. WordPress core itself (without any `object-cache.php` drop-in) caches many things, including e.g. theme mods within a single request, and that does not interfere with theme development. So what is happening here is not different from that.

> As I mentioned in the [other comment](https://github.com/WordPress/wordpress-develop/pull/3712/files#r1071617204), seems WP needs a way (another constant?) to signify "development mode". Once that is added, these uses of `WP_DEBUG` can be switched over/removed.

Agreed. But per the above, I don't think we need to rush the half-baked solution of `WP_DEBUG` into core here, especially since it is not necessary for non-persistent caches like this. Which leads me back to something like the two options I was referring to in https://github.com/WordPress/wordpress-develop/pull/3556#discussion_r1071510762: I would advise that we remove the `WP_DEBUG` check from this PR since it's not needed, and open a separate Trac ticket to explore a way to signify "development mode". For any theme caches that are persistent, we would need to use that then.

Alternatively we build that "development mode" thing into this PR, but I think that would be unnecessarily rushing it, and again, not necessary.

Reply:
Building a "development mode" solution is out of scope for this PR. I'd suggest that work be done separately in a separate Trac ticket. Why? A development mode strategy can be larger that this specific smaller chunk of it.

Once that strategy is implemented and committed, then it can replace the `WP_DEBUG` here and anywhere else that is using this constant to denote "development mode".

Reply:
>But per the above, I don't think we need to rush the half-baked solution of WP_DEBUG into core here, especially since it is not necessary for non-persistent caches like this. 

@felixarntz some questions please:

- Why is `WP_DEBUG` half-baked? 
- What are the risks? 
- Is there _any_ chance that not having it could disrupt a developer's workflow?

Reply:
@hellofromtonya Agreed that the "development mode" is out of scope for this PR.

Replying to your questions:
* `WP_DEBUG` is "half-baked" IMO because it is a workaround for the lack of having a "development mode".
* The risks of keeping it as is are that other (core) developers may see this and consider it a good practice to use for some other part of core as well (e.g. in some other unrelated PR that realistically we may potentially miss). It is a reality that any code in core may be used as a reference (a.k.a. "core already does this here"), and I would want to make sure this is not the case with this usage of `WP_DEBUG`.
* I do not see any risk for removing it due to the caching being non-persistent. When a site is not using an object cache, all caches, including other ones that are related to theme development (e.g. theme mods) are non-persistent, and we don't check for `WP_DEBUG` in those situations either.

My preference remains to leave the `WP_DEBUG` condition out of it, but I realize there are opposing voices. So the least I would ask for if we keep it would be to annotate it (also in the other PRs where we do this, like #3712 for example) with some kind of `TODO` comment annotation to clarify that this should be replaced with a development mode check in the near future.

Is the latter a reasonable compromise?

Reply:
>So the least I would ask for if we keep it would be to annotate it (also in the other PRs where we do this, like https://github.com/WordPress/wordpress-develop/pull/3712 for example) with some kind of TODO comment annotation to clarify that this should be replaced with a development mode check in the near future.
>Is the latter a reasonable compromise?

I think that's a great idea 💡 👏  Let me add it for your review.

Reply:
@felixarntz here's the `@todo` note commit https://github.com/WordPress/wordpress-develop/pull/3556/commits/c4a20376f3c8146a5273fba3e6105d054bc3b426. What do you think?

Reply:
@hellofromtonya That's great, thanks!